File: jetbrains-plugin/src/test/kotlin/com/mariadbprofiler/plugin/model/JobInfoTest.kt
package com.mariadbprofiler.plugin.model

import kotlinx.serialization.json.Json
import org.junit.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue

class JobInfoTest {

    private val json = Json { ignoreUnknownKeys = true; isLenient = true }

    @Test
    fun `active job has no end time`() {
        val job = JobInfo(key = "test-key", startedAt = 1705970401.0)
        assertTrue(job.isActive)
        assertNull(job.endedAt)
        assertEquals("running...", job.formattedDuration)
    }

    @Test
    fun `completed job has end time and duration`() {
        val job = JobInfo(
            key = "test-key",
            startedAt = 1705970401.0,
            endedAt = 1705970411.0,
            queryCount = 42
        )
        assertFalse(job.isActive)
        assertEquals(10.0, job.durationSeconds)
        assertEquals("10.0 s", job.formattedDuration)
    }

    @Test
    fun `short duration displays as milliseconds`() {
        val job = JobInfo(
            key = "test",
            startedAt = 1705970401.0,
            endedAt = 1705970401.5
        )
        assertEquals("500 ms", job.formattedDuration)
    }

    @Test
    fun `long duration displays as minutes`() {
        val job = JobInfo(
            key = "test",
            startedAt = 1705970401.0,
            endedAt = 1705970581.0 // 180 seconds
        )
        assertEquals("3.0 min", job.formattedDuration)
    }

    @Test
    fun `parse jobs file`() {
        val content = """
        {
            "active_jobs": {
                "abc123": {"started_at": 1705970401.0, "parent": null}
            },
            "completed_jobs": {
                "def456": {"started_at": 1705970300.0, "ended_at": 1705970401.0, "query_count": 42, "parent": null}
            }
        }
        """.trimIndent()

        val jobsFile = json.decodeFromString<JobsFile>(content)
        assertEquals(1, jobsFile.activeJobs.size)
        assertEquals(1, jobsFile.completedJobs.size)
        assertEquals(1705970401.0, jobsFile.activeJobs["abc123"]?.startedAt)
        assertEquals(42, jobsFile.completedJobs["def456"]?.queryCount)
    }
}

File: vscode-extension/test/unit/JobInfo.test.ts
import { describe, it, expect } from 'vitest';
import {
  parseJobsFile,
  jobsFileToJobInfos,
  formatDuration,
} from '../../src/model/JobInfo';

describe('parseJobsFile', () => {
  it('should parse standard jobs.json', () => {
    const json = JSON.stringify({
      active_jobs: {
        'abc-123': { started_at: 1705970401.0, parent: null },
      },
      completed_jobs: {
        'def-456': { started_at: 1705970300.0, ended_at: 1705970400.0, query_count: 42 },
      },
    });

    const result = parseJobsFile(json);
    expect(Object.keys(result.active_jobs)).toEqual(['abc-123']);
    expect(Object.keys(result.completed_jobs)).toEqual(['def-456']);
    expect(result.active_jobs['abc-123'].started_at).toBe(1705970401.0);
    expect(result.completed_jobs['def-456'].query_count).toBe(42);
  });

  it('should handle PHP empty array as empty map', () => {
    const json = JSON.stringify({
      active_jobs: [],
      completed_jobs: { 'def-456': { started_at: 1705970300.0 } },
    });

    const result = parseJobsFile(json);
    expect(result.active_jobs).toEqual({});
    expect(Object.keys(result.completed_jobs)).toEqual(['def-456']);
  });

  it('should handle both maps empty (PHP format)', () => {
    const json = JSON.stringify({
      active_jobs: [],
      completed_jobs: [],
    });

    const result = parseJobsFile(json);
    expect(result.active_jobs).toEqual({});
    expect(result.completed_jobs).toEqual({});
  });

  it('should handle parent field', () => {
    const json = JSON.stringify({
      active_jobs: {},
      completed_jobs: {
        'child-job': { started_at: 100, ended_at: 200, query_count: 5, parent: 'parent-job' },
      },
    });

    const result = parseJobsFile(json);
    expect(result.completed_jobs['child-job'].parent).toBe('parent-job');
  });
});

describe('jobsFileToJobInfos', () => {
  it('should convert to sorted JobInfo array', () => {
    const json = JSON.stringify({
      active_jobs: {
        'active-1': { started_at: 100 },
      },
      completed_jobs: {
        'completed-1': { started_at: 50, ended_at: 80, query_count: 10 },
        'completed-2': { started_at: 90, ended_at: 95, query_count: 5 },
      },
    });

    const jobsFile = parseJobsFile(json);
    const infos = jobsFileToJobInfos(jobsFile);

    // Active jobs first
    expect(infos[0].key).toBe('active-1');
    expect(infos[0].isActive).toBe(true);

    // Completed sorted by startedAt descending
    expect(infos[1].key).toBe('completed-2');
    expect(infos[1].isActive).toBe(false);
    expect(infos[2].key).toBe('completed-1');
    expect(infos[2].isActive).toBe(false);
    expect(infos[2].queryCount).toBe(10);
  });

  it('should handle empty jobs file', () => {
    const jobsFile = parseJobsFile(JSON.stringify({ active_jobs: [], completed_jobs: [] }));
    expect(jobsFileToJobInfos(jobsFile)).toEqual([]);
  });
});

describe('formatDuration', () => {
  it('should format short duration in seconds', () => {
    expect(formatDuration(100, 103.2)).toBe('3.2s');
  });

  it('should format longer duration with minutes', () => {
    expect(formatDuration(100, 225)).toBe('2m5s');
  });

  it('should handle zero duration', () => {
    expect(formatDuration(100, 100)).toBe('0.0s');
  });
});
